kirill_tolstobrov/main.cpp

```diff
@@ -143,49 +207,136 @@ class Graph {
+            depths_map.push_back({vertices_amount - 1, vertices_amount - 1});
+          } else {
+            depths_map[i + 1].second++;
+          }
+        }
+      }
+    }
+
+    new_vertex_prob -= probability_decreasement;
   }
-  for (const auto& connection : connections) {
-    graph.bind_vertices(connection.first, connection.second);
+  return vertices_amount;
+}
+
+void generate_yellow_edge(Graph& graph,
+                          std::vector<std::pair<int, int>>& depths_map,
+                          int cur_depth,
+                          int cur_id) {
+  int next_depth_start = depths_map[cur_depth + 1].first;
+  int next_depth_finish = depths_map[cur_depth + 1].second;
+
+  int binded_id = next_depth_start +
+                  (std::rand() % (next_depth_finish - next_depth_start + 1));
+
+  graph.bind_vertices(cur_id, binded_id, Color::Yellow);
+}
+
+void generate_red_edge(Graph& graph,
+                       std::vector<std::pair<int, int>>& depths_map,
+                       int cur_depth,
+                       int cur_id) {
+  int next_next_depth_start = depths_map[cur_depth + 2].first;
+  int next_next_depth_finish = depths_map[cur_depth + 2].second;
+
+  int binded_id =
+      next_next_depth_start +
+      (std::rand() % (next_next_depth_finish - next_next_depth_start + 1));
+
+  graph.bind_vertices(cur_id, binded_id, Color::Red);
+}
+
+Graph generate_random_graph(int depth, int new_vertices_num) {
+  Graph graph = Graph();
+
+  srand(std::time(NULL));
+
+  const float probability_decreasement = 1.0 / depth;
+  graph.add_new_vertex();
+  // depths_map: depth -> (first_vertex_id; last_vertex_id)
+  std::vector<std::pair<int, int>> depths_map;
+  depths_map.push_back({0, 0});
+
+  int vertices_amount = generate_grey_edges(
+      graph, depth, new_vertices_num, depths_map, probability_decreasement);
+
+  enum { GREEN_PROB = 10, BLUE_PROB = 25, RED_PROB = 33, PROB_COEF = 100 };
+  int cur_depth = 0;
+  float yellow_probability = 0;
+  const float probability_increasement = probability_decreasement;
+
+  for (int cur_id = 0; cur_id < vertices_amount; cur_id++) {
+    if (random_prob() <= GREEN_PROB / (float)PROB_COEF) {
+      graph.bind_vertices(cur_id, cur_id, Color::Green);
+    }
+
+    if (cur_id != depths_map[cur_depth].second) {
+      if (random_prob() <= BLUE_PROB / (float)PROB_COEF) {
+        graph.bind_vertices(cur_id, cur_id + 1, Color::Blue);
+      }
+    }
+
+    if (cur_depth != depth &&
+        random_prob() <= yellow_probability * 100 / (float)PROB_COEF) {
+      generate_yellow_edge(graph, depths_map, cur_depth, cur_id);
+    }
+
+    if (cur_depth < (depth - 1) &&
+        random_prob() <= RED_PROB / (float)PROB_COEF) {
+      generate_red_edge(graph, depths_map, cur_depth, cur_id);
+    }
+
+    if (cur_id == depths_map[cur_depth].second) {
+      cur_depth++;
+      yellow_probability += probability_increasement;
+    }
   }
+
+  graph.set_depths_map(depths_map);
   return graph;
 }
```

Comment:
Очень длинная функция, лучше разбить на более мелкие, примерно:
```suggestion
Graph generate_random_graph(int depth, int new_vertices_num) {
  auto graph = Graph();
  generate_grey_edges(graph, depth, new_vertices_num);
  generate_green_edges(graph);
  generate_blue_edges(graph);
  generate_yellow_edges(graph);
  generate_red_edges(graph);
  return graph;
}
```